docs/spec_draft.md

```diff
@@ -1917,6 +1918,70 @@ and produces a `result` tensor. More formally,
 ```
 
 [Back to Ops](#index-of-ops)
+
+## stablehlo.rng_bit_generator
+
+### Semantics
+
+Returns an `output` filled with uniform random bits and an updated output state
+`output_state` given an initial state `initial_state` using the pseudorandom
+number generator algorithm `rng_algorithm`. The output is guaranteed to be
+deterministic function of `initial_state`, but it is not guaranteed to be
+deterministic between backends and different compiler versions.
+
+`rng_algorithm` is one of the following, different variants of the algorithm is
+implementation-defined:
+  * `DEFAULT`: Backend specific algorithm.
+  * `THREE_FRY`: Threefry counter-based PRNG algorithm.*
+  * `PHILOX`: Philox algorithm to generate random numbers in parallel.*
+
+\* See: [Salmon et al. SC 2011. Parallel random numbers: as easy as 1, 2, 3.
+](http://www.thesalmons.org/john/random123/papers/random123sc11.pdf)
+
+### Inputs
+
+| Name            | Type                                         |
+|-----------------|----------------------------------------------|
+| `initial_state` | tensor of integer or floating-point type     |
+| `rng_algorithm` | enum of `DEFAULT`, `THREE_FRY`, and `PHILOX` |
+
+### Outputs
+
+| Name           | Type                                     |
+|----------------|------------------------------------------|
+| `output_state` | tensor of integer or floating-point type |
+| `output`       | tensor of integer or floating-point type |
+
+### Constraints
+
+  * (C1) shape(`initial_state`) $=$ shape(`output_state`).
+  * (C2) For `initial_state`, `output_state`, and `output`:
+    * If `rng_algorithm = DEFAULT`:
+      * `initial_state`, `output_state`, and `output` have backend-specific type
+        requirements.
+    * If `rng_algorithm = THREE_FRY`:
+      * type(`initial_state`) $=$ type(`output_state`) $=$ `tensor<2xui64>`.
+      * element_type(`output`) $\in$ {`ui32`, `ui64`}.
+    * If `rng_algorithm = PHILOX`:
+      * type(`initial_state`) $=$ type(`output_state`) $=$ `tensor<3xui64>`.
+      * element_type(`output`) $\in$ {`ui32`, `ui64`}.
```

Comment:
Proposed rewording:

```
  * (C1) `type(initial_state) = type(output_state) = tensor<2xui64>`.
  <the end, no more constraints>
```

Feedback:
  1) Per [hlo_verifier.cc](http://google3/third_party/tensorflow/compiler/xla/service/hlo_verifier.cc;l=928-929;rcl=487909736), I think we can make a stronger claim that `type(initial_state) = type(output_state)`.
  2) Furthermore, according to [lax.py](https://github.com/google/jax/blob/40e81c3a8655709c9aa554b52e1a85842786e034/jax/_src/lax/lax.py#L4373-L4380), both `initial_state` and `output_state` are always `tensor<2xui64>`?
  3) Output's element_type can be covered in Inputs/Outputs.

Reply:
I separated into these constraints since I do not know what `DEFAULT` is capable of doing. It can possibly have different shape and element type.

Reply:
Since DEFAULT is open ended, it makes sense to have the current proposed. However, we could use your proposed rewording and note this in the same issue.

Reply:
Let's bias towards the simplest formulation (i.e the one proposed above) and sort this out in the ticket.

Reply:
The proposed rewording is missing constraint for `output`, and PHILOX algorithm requires `tensor<3xui64>` whereas THREE_FRY is `tensor<2xui64>`. How about:
```
  * (C1) If `rng_algorithm` is:
    * `THREE_FRY`: `type(initial_state) = type(output_state) = tensor<2xui64>`.
    * `PHILOX`: `type(initial_state) = type(output_state) = tensor<3xui64>`.
  * (C2) element_type(`output`) $\in$ {`ui32`, `ui64`}.
```

Reply:
PHILOX (at least on XLA CPU) is fine with both `2xui64` and `3xui64` (`2xui64` is important because this is what JAX produces). Perhaps we can use the following format to simplify speccing this:

```
  * (C1) `type(initial_state) = type(output_state)`.
  * (C2) `size(initial_state)` depends on `rng_algorithm`:
    * `DEFAULT`: implementation-defined.
    * `THREE_FRY`: `2`.
    * `PHILOX`: `2` or `3`.
```

For the element type, where is the restriction coming from? [xla_builder.cc](https://github.com/tensorflow/tensorflow/blob/8b7373a66dfcb8ccf994ea0b1d720b34a4d32cbf/tensorflow/compiler/xla/client/xla_builder.cc#L2566-L2601) happily admits all floating-point and integer types. [hlo_verifier.cc](https://github.com/tensorflow/tensorflow/blob/8b7373a66dfcb8ccf994ea0b1d720b34a4d32cbf/tensorflow/compiler/xla/service/hlo_verifier.cc#L920-L938) doesn't have an opinion. I would imagine that we don't need to cover this in Constraints?

Reply:
Thanks for catching this; I wasn't aware that PHILOX could take `2xui64` in [JAX](https://github.com/google/jax/blob/40e81c3a8655709c9aa554b52e1a85842786e034/jax/_src/lax/lax.py#L4373-L4380) because [XLA doc](https://www.tensorflow.org/xla/operation_semantics#rngbitgenerator) showed otherwise.

Since StableHLO supports 4-bit integers while [xla_builder.cc](https://github.com/tensorflow/tensorflow/blob/8b7373a66dfcb8ccf994ea0b1d720b34a4d32cbf/tensorflow/compiler/xla/client/xla_builder.cc#L2566-L2601) does not, and [hlo_verifier.cc](https://github.com/tensorflow/tensorflow/blob/8b7373a66dfcb8ccf994ea0b1d720b34a4d32cbf/tensorflow/compiler/xla/service/hlo_verifier.cc#L920-L938) also does not have verification for supported element types, I believe we have to call out exactly which element types are supported regardless, either in Inputs section or in Constraints. I'm leaning towards the Constraints, which may look like below. Suggestions?

```
  * (C1) type(`initial_state`) $=$ type(`output_state`).
  * (C2) size(`initial_state`) depends on `rng_algorithm`:
    * `DEFAULT`: implementation-defined.
    * `THREE_FRY`: `2`.
    * `PHILOX`: `2` or `3`.
  * (C3) element_type(`initial_state`) depends on `rng_algorithm`:
    * `DEFAULT`: implementation-defined.
    * `THREE_FRY`: `ui8`, `ui16`, `ui32`, `ui64`, `si8`, `si16`, `si32`, `si64`,
      `bf16`, `f16`, `f32`, or `f64`.
    * `PHILOX`: `ui8`, `ui16`, `ui32`, `ui64`, `si8`, `si16`, `si32`, `si64`,
      `bf16`, `f16`, `f32`, or `f64`.
```

Reply:
The situation with 4-bit integers is not unique to this op. E.g. `stablehlo.add` supports them, while `kAdd` doesn't. I wouldn't add special treatment for si4/ui4 here in this specific op. Opened #661 to track the general question.